### PR TITLE
Add Escargot-Debugger test runner

### PR DESCRIPTION
**Pull request by kisbg**
To use the Escargot Debugger test system.
First the Escargot needs to be build with the -DESCARGOT_DEBUGGER=1 option, then
run the ./tools/run-test.py with the  escargot-debugger command

Signed-off-by: bence gabor kis <kisbg@inf.u-szeged.hu>

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -58,12 +58,12 @@ def isPr() {
...
-                    sh 'LDFLAGS=" -L/usr/icu32/lib/  -Wl,-rpath=/usr/icu32/lib/" cmake  -H./ -Bbuild/out_linux -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x86 -DESCARGOT_MODE=debug -DESCARGOT_OUTPUT=shell_test -GNinja'
+                    sh 'LDFLAGS=" -L/usr/icu32/lib/  -Wl,-rpath=/usr/icu32/lib/" cmake  -H./ -Bbuild/out_linux -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x86 -DESCARGOT_MODE=debug -DESCARGOT_DEBUGGER=1 -DESCARGOT_OUTPUT=shell_test -GNinja'
                     sh 'LDFLAGS=" -L/usr/icu32/lib/  -Wl,-rpath=/usr/icu32/lib/" cmake  -H./ -Bbuild/out_linux_release -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x86 -DESCARGOT_MODE=release -DESCARGOT_OUTPUT=shell_test -GNinja'
                     sh 'gcc -shared -m32 -fPIC -o backtrace-hooking-32.so tools/test/test262/backtrace-hooking.c'
                 },
                 '64bit' : {
-                    sh 'cmake  -H./ -Bbuild/out_linux64 -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x64 -DESCARGOT_MODE=debug -DESCARGOT_OUTPUT=shell_test -GNinja'
+                    sh 'cmake  -H./ -Bbuild/out_linux64 -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x64 -DESCARGOT_MODE=debug -DESCARGOT_DEBUGGER=1 -DESCARGOT_OUTPUT=shell_test -GNinja'
```
Comment (zherczeg) [inline Jenkinsfile]: Shouldn't we create another binary rather than changing an existing one?

**Comment by zherczeg** [comment]
It would be good to add some actual tests for existing features (step/next/continue)

File: Jenkinsfile (hunk 1 above)
Comment (kisbg) [reply Jenkinsfile]: I didn't want to add new binary, since it doesn't influence the other test-cases.

**Comment by kisbg** [comment]
@zherczeg  Of course, but first i wanted to test the new Jenkins settings.

File: tools/debugger/debugger_tester.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+START_DEBUG_SERVER="${ESCARGOT} --start-debug-server ${TEST_CASE} &"
+three=3
+
+echo "$START_DEBUG_SERVER"
+eval "$START_DEBUG_SERVER"
+sleep 2s
+
+RESULT_TEMP=`mktemp ${TEST_CASE}.out.XXXXXXXXXX`
+
+(cat "${COMMANDS}" | ${DEBUGGER_CLIENT}  --display  ${three}) >${RESULT_TEMP} 2>&1
```
Comment (zherczeg) [inline tools/debugger/debugger_tester.sh]: I am not sure `--display 3` needs to be enabled by default. If needed, it can be enabled in the test.

File: tools/debugger/debugger_core.py [hunk 3]
```diff
@@ -256,10 +256,10 @@ def __del__(self):
             self.channel.close()
 
     def encode8(self, string):
-        return string.encode("latin1")
+        return string.decode("latin1")
 
     def encode16(self, string):
-        return string.encode("UTF-16LE" if self.little_endian else "UTF-16BE", "namereplace")
+        return string.decode("UTF-16LE" if self.little_endian else "UTF-16BE", "namereplace")
```
Comment (zherczeg) [inline tools/debugger/debugger_core.py]: Rename the function (encode16 -> decode16). Also for encode8.

File: tools/debugger/debugger_tester.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+sleep 1s
+
+RESULT_TEMP=`mktemp ${TEST_CASE}.out.XXXXXXXXXX`
+
+cat "${COMMANDS}" | ${DEBUGGER_CLIENT}  --non-interactive >${RESULT_TEMP} 2>&1
+
+# if [[ $TEST_CASE == *"restart"* ]]; then
+#   CONTINUE_CASE=$(sed "s/restart/continue/g" <<< "$TEST_CASE")
+#   (cat "${CONTINUE_CASE}.cmd" | ${DEBUGGER_CLIENT} --channel ${CHANNEL} --non-interactive ${CLIENT_ARGS}) >>${RESULT_TEMP} 2>&1
+# fi
```
Comment (zherczeg) [inline tools/debugger/debugger_tester.sh]: What is this commented out part?

File: Jenkinsfile (hunk 1 above)
Comment (clover2123) [reply Jenkinsfile]: Currently CI(Jenkins) runs `debug binary` for modifiedVendorTest, regression-tests, new-es, intl, sunspider-js and test262 TCs.
These tests take quite a short time, but test262 case might be affected by running `debugger-enabled binary` since test262 is one of the most time-consuming TCs.

Comment (clover2123) [reply Jenkinsfile]: I'm not sure but if the total time is significantly increased it would be better to run `debug-only binary` for test262 case. What do you think about it?

Comment (kisbg) [reply Jenkinsfile]: I added a debugger build to Jenkins. The only question have do we want to have both a 32-bit and a 64-bit debugger build or it is a 64-bit build enough?

Comment (clover2123) [reply Jenkinsfile]: ATM 64-bit build would be enough. We may add 32-bit build later if it's really necessary. Thanks.

File: Jenkinsfile [hunk 5]
```diff
@@ -58,14 +58,18 @@ def isPr() {
             stage('Prepare build(gcc)') {
             parallel (
                 '32bit' : {
-                    sh 'LDFLAGS=" -L/usr/icu32/lib/  -Wl,-rpath=/usr/icu32/lib/" cmake  -H./ -Bbuild/out_linux -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x86 -DESCARGOT_MODE=debug -DESCARGOT_OUTPUT=shell_test -GNinja'
+                    sh 'LDFLAGS=" -L/usr/icu32/lib/  -Wl,-rpath=/usr/icu32/lib/" cmake  -H./ -Bbuild/out_linux -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x86 -DESCARGOT_MODE=debug  -DESCARGOT_OUTPUT=shell_test -GNinja'
```
Comment (zherczeg) [inline Jenkinsfile]: Revert this change. Space is not needed.

File: Jenkinsfile [hunk 6]
```diff
@@ -58,14 +58,18 @@ def isPr() {
...
                     sh 'LDFLAGS=" -L/usr/icu32/lib/  -Wl,-rpath=/usr/icu32/lib/" cmake  -H./ -Bbuild/out_linux_release -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x86 -DESCARGOT_MODE=release -DESCARGOT_OUTPUT=shell_test -GNinja'
                     sh 'gcc -shared -m32 -fPIC -o backtrace-hooking-32.so tools/test/test262/backtrace-hooking.c'
                 },
                 '64bit' : {
-                    sh 'cmake  -H./ -Bbuild/out_linux64 -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x64 -DESCARGOT_MODE=debug -DESCARGOT_OUTPUT=shell_test -GNinja'
+                    sh 'cmake  -H./ -Bbuild/out_linux64 -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x64 -DESCARGOT_MODE=debug  -DESCARGOT_OUTPUT=shell_test -GNinja'
```
Comment (zherczeg) [inline Jenkinsfile]: Ditto.

File: Jenkinsfile [hunk 7]
```diff
@@ -58,14 +58,18 @@ def isPr() {
...
                     sh 'cmake  -H./ -Bbuild/out_linux64_release -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x64 -DESCARGOT_MODE=release -DESCARGOT_OUTPUT=shell_test -GNinja'
                     sh 'gcc -shared -fPIC -o backtrace-hooking-64.so tools/test/test262/backtrace-hooking.c'
+                },
+                'debugger-binary' : {
+                  sh 'cmake  -H./ -Bbuild/debugger_out_linux64 -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x64 -DESCARGOT_MODE=debug  -DESCARGOT_DEBUGGER=1 -DESCARGOT_OUTPUT=shell_test -GNinja'
+
```
Comment (zherczeg) [inline Jenkinsfile]: Extra newline.

File: Jenkinsfile [hunk 8]
```diff
@@ -122,6 +127,10 @@ def isPr() {
                         sh 'tools/run-tests.py --arch=x86_64 --engine="${WORKSPACE}/build/out_linux64_release/escargot" jetstream-only-cdjs modifiedVendorTest jsc-stress sunspider-js'
                         sh 'tools/run-tests.py --arch=x86_64 --engine="${WORKSPACE}/build/out_linux64_release/escargot" v8 spidermonkey regression-tests new-es intl'
                     },
+                    'Escargot-debugger-test-64bit' : {
+                        sh 'tools/run-tests.py --arch=x86_64 --engine="${WORKSPACE}/build/debugger_out_linux64/escargot" escargot-debugger'
+                    },
+
```
Comment (zherczeg) [inline Jenkinsfile]: Extra newline.

File: tools/debugger/debugger_tester.sh [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+ESCARGOT=$1
+TEST_CASE=$2
+COMMANDS=$3
+EXPECTED=$4
+DEBUGGER_CLIENT=$5
+START_DEBUG_SERVER="${ESCARGOT} --start-debug-server ${TEST_CASE} &"
+eval "$START_DEBUG_SERVER"
```
Comment (zherczeg) [inline tools/debugger/debugger_tester.sh]: I think you can directly evaluate the command here, no need any `eval`.

File: tools/debugger/tests/do_break.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
+// Copyright JS Foundation and other contributors, http://js.foundation
```
Comment (zherczeg) [inline tools/debugger/tests/do_break.js]: Wrong license. Fix all.

File: tools/run-tests.py [hunk 11]
```diff
@@ -503,6 +504,31 @@ def run_intl(engine, arch):
...
+    Escargot_Debugger_Test_DIR = join(PROJECT_SOURCE_DIR, 'tools', 'debugger', 'tests')
+    Escargot_Debugger_Client = join(PROJECT_SOURCE_DIR, 'tools', 'debugger', 'debugger.py')
+    Escargot_Debugger_Tester = join(PROJECT_SOURCE_DIR, 'tools', 'debugger', 'debugger_tester.sh')
+    print('Running Escargot-debugger test:')
+    test_cases = glob(join(Escargot_Debugger_Test_DIR, '*.js'))
+    test_commands = glob(join(Escargot_Debugger_Test_DIR, '*.cmd'))
+    expected_outputs = glob(join(Escargot_Debugger_Test_DIR, '*.expected'))
+    test_cases.sort()
+    test_commands.sort()
+    expected_outputs.sort()
```
Comment (zherczeg) [inline tools/run-tests.py]: We have three sorted sets, but what ensures, that we get the same tests in all (e.g. if one expected is missing, we get wrong order)? I would only use one (`*.cmd`), and use string replace to figure out the corresponding name (e.g. `.cmd` -> `.js`).

File: tools/debugger/tests/do_break.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
+ *
```
Comment (clover2123) [inline tools/debugger/tests/do_break.js]: Incorrect year at here. 2016 should be fixed to 2020.

File: tools/debugger/tests/do_continue.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
+ *
```
Comment (clover2123) [inline tools/debugger/tests/do_continue.js]: ditto.

File: tools/debugger/tests/do_step.js [hunk 14]
```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2016-present Samsung Electronics Co., Ltd
+ *
```
Comment (clover2123) [inline tools/debugger/tests/do_step.js]: Same as above.

File: tools/run-tests.py [hunk 15]
```diff
@@ -503,6 +504,29 @@ def run_intl(engine, arch):
     if fails > 0:
         raise Exception('Intl tests failed')
 
+@runner('escargot-debugger', default=True)
+def run_escargot_debugger(engine, arch):
+    Escargot_Debugger_Test_DIR = join(PROJECT_SOURCE_DIR, 'tools', 'debugger', 'tests')
+    Escargot_Debugger_Client = join(PROJECT_SOURCE_DIR, 'tools', 'debugger', 'debugger.py')
+    Escargot_Debugger_Tester = join(PROJECT_SOURCE_DIR, 'tools', 'debugger', 'debugger_tester.sh')
```
Comment (clover2123) [inline tools/run-tests.py]: It's a minor suggestion. 
How about using uppercase for these variable names since all other variables are declared by uppercase names in this script file.

File: tools/debugger/debugger_tester.sh [hunk 16]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+
+ESCARGOT=$1
+TEST_CASE=$2
+DEBUGGER_CLIENT=$3
+START_DEBUG_SERVER="${ESCARGOT} --start-debug-server ${TEST_CASE}.js &"
+eval ${START_DEBUG_SERVER}
```
Comment (zherczeg) [inline tools/debugger/debugger_tester.sh]: I think you can evaluate `${ESCARGOT} --start-debug-server ${TEST_CASE}.js &` directly here, you don't need an 'eval'.

**Comment by zherczeg** [comment]
Do the changes in ` third_party/GCutil`needed?

**Comment by zherczeg** [comment]
The extra sapces are still in the patch. Are they needed?

**Comment by kisbg** [comment]
@zherczeg  I check the pull request, but i don't see any unnecessary new line. Can you tag a unnecessary empty line? Thank you, in advance!

File: Jenkinsfile [hunk 17]
```diff
@@ -58,14 +58,17 @@ def isPr() {
             stage('Prepare build(gcc)') {
             parallel (
                 '32bit' : {
-                    sh 'LDFLAGS=" -L/usr/icu32/lib/  -Wl,-rpath=/usr/icu32/lib/" cmake  -H./ -Bbuild/out_linux -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x86 -DESCARGOT_MODE=debug -DESCARGOT_OUTPUT=shell_test -GNinja'
+                    sh 'LDFLAGS=" -L/usr/icu32/lib/  -Wl,-rpath=/usr/icu32/lib/" cmake  -H./ -Bbuild/out_linux -DESCARGOT_HOST=linux -DESCARGOT_ARCH=x86 -DESCARGOT_MODE=debug  -DESCARGOT_OUTPUT=shell_test -GNinja'
```
Comment (zherczeg) [inline Jenkinsfile]: Here is the unnecessary space.

Comment (kisbg) [reply Jenkinsfile]: thank you!
